src/matrices.jl

```diff
@@ -0,0 +1,108 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENCE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    JMatrix{T}
+
+Square matrix of ones.
+"""
+struct JMatrix{T} end
+(J::JMatrix{T})(d::Integer) where {T} = Ones{T}(d, d)
+*(::JMatrix, v::AbstractVector) = Fill(sum(v), length(v))
+*(v::Adjoint{<:Any, <:AbstractVector}, J::JMatrix) = (J*v')'
+
+"""
+    FMatrix{T}
+
+`F` matrix, as defined by Aitchison 1986.
+"""
+struct FMatrix{T} end
+(F::FMatrix{T})(d::Integer) where {T} = [Diagonal(fill(one(T), d)) -Ones{T}(d)]
+*(::FMatrix{T}, v::AbstractVector) where {T} = Fill(-last(v), length(v)-1) + v[1:end-1]
```

Comment:
I think in this case we don't gain much by using `Fill`? Can we simplify it to `v[1:end-1] .- v[end]`?

Reply:
you're right: for `v = [1:1000000...];`, both f (old method) and g (new method) run fine.

```julia> @benchmark f(v)
BenchmarkTools.Trial: 3820 samples with 1 evaluation.
 Range (min … max):  1.205 ms …   1.995 ms  ┊ GC (min … max): 0.00% … 27.04%
 Time  (median):     1.274 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.305 ms ± 103.803 μs  ┊ GC (mean ± σ):  1.30% ±  4.94%

    ▁▅▇██▇▆▆▅▄▃▃▂▂▁▂▁▁▁ ▁                                     ▂
  ███████████████████████▇▇▇█▆▇▆▅▆▅▃▄▄▃▄▁▃▄▃▁▁▁▁▆█▅▅▁▃▃▅███▇▇ █
  1.21 ms      Histogram: log(frequency) by time       1.8 ms <

 Memory estimate: 15.26 MiB, allocs estimate: 4.

julia> @benchmark g(v)
BenchmarkTools.Trial: 2930 samples with 1 evaluation.
 Range (min … max):  1.258 ms … 6.590 ms  ┊ GC (min … max): 0.00% … 78.81%
 Time  (median):     1.330 ms             ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.703 ms ± 1.035 ms  ┊ GC (mean ± σ):  5.96% ± 10.54%

  █▆▂                        ▁ ▃       ▁
  ███▇▁▁▁▁▁▁▁▁▁▃▃▁▁▁▁▁▁▁▁▁▁█████▅▃▁▁▃▃▆█▁▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ █
  1.26 ms     Histogram: log(frequency) by time     6.46 ms <

 Memory estimate: 15.26 MiB, allocs estimate: 4.
```

Reply:
Yes, both are fast and in place operations:

```julia> @benchmark oldF(v)
BenchmarkTools.Trial: 4151 samples with 1 evaluation.
 Range (min … max):  971.105 μs …   5.308 ms  ┊ GC (min … max):  0.00% … 16.27%
 Time  (median):       1.033 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.201 ms ± 279.975 μs  ┊ GC (mean ± σ):  10.69% ± 11.69%

   ▇▁▁█▂                                        ▂
  ▃█████▅▃▃▃▂▂▂▂▂▂▂▂▁▂▂▂▂▂▂▂▂▂▁▁▁▁▂▂▁▁▁▁▁▁▁▁▁▁▃██▇▅▃▃▃▂▂▂▂▂▂▂▂▂ ▃
  971 μs           Histogram: frequency by time          1.7 ms <

 Memory estimate: 15.26 MiB, allocs estimate: 4.

julia> @benchmark new(v)
BenchmarkTools.Trial: 2509 samples with 1 evaluation.
 Range (min … max):  925.259 μs …   5.097 ms  ┊ GC (min … max):  0.00% … 17.32%
 Time  (median):       1.525 ms               ┊ GC (median):     0.00%
 Time  (mean ± σ):     1.990 ms ± 884.611 μs  ┊ GC (mean ± σ):  11.55% ± 16.42%

       ▃█▆▂                ▇▅▂▁                         ▅▄▂▁    ▁
  ▃▁▁▄▇█████▇▃▄▃▁▁▁▁▁▁▁▁▁▁██████▇▅▆▆▅▃▁▃▄▁▃▁▁▁▁▁▃▃▁▃▁▃▁▅█████▅▇ █
  925 μs        Histogram: log(frequency) by time       3.94 ms <

 Memory estimate: 15.26 MiB, allocs estimate: 4.
```

Reply:
Done. Also used `I(d)` instead of `Diagonal` in the constructor.